pkg/messaging/rabbitmq/pubsub.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package rabbitmq
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	log "github.com/mainflux/mainflux/logger"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const (
+	chansPrefix = "channels"
+	// SubjectAllChannels represents subject to subscribe for all the channels.
+	SubjectAllChannels = "channels.>"
+	exchangeName       = "mainflux-exchange"
+	exchangeKind       = "direct"
+)
+
+var (
+	ErrAlreadySubscribed = errors.New("already subscribed to topic")
+	ErrNotSubscribed     = errors.New("not subscribed")
+	ErrEmptyTopic        = errors.New("empty topic")
+	ErrEmptyID           = errors.New("empty id")
+)
+
+var _ messaging.PubSub = (*pubsub)(nil)
+
+// PubSub wraps messaging Publisher exposing
+// Close() method for RabbitMQ connection.
+type PubSub interface {
+	messaging.PubSub
+	Close()
+}
+
+type subscription struct {
+	cancel func() error
+}
+type pubsub struct {
+	publisher
+	queue         string
+	logger        log.Logger
+	subscriptions map[string]map[string]subscription
+	mu            sync.Mutex
+}
+
+// NewPubSub returns RabbitMQ message publisher/subscriber.
+func NewPubSub(url, queue string, logger log.Logger) (PubSub, error) {
+	endpoint := fmt.Sprintf("amqp://%s", url)
+	conn, err := amqp.Dial(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	ch, err := conn.Channel()
+	if err != nil {
+		return nil, err
+	}
+	if err := ch.ExchangeDeclare(exchangeName, amqp.ExchangeDirect, true, false, false, false, nil); err != nil {
+		return nil, err
+	}
+	ret := &pubsub{
+		publisher: publisher{
+			conn: conn,
+			ch:   ch,
+		},
+		logger:        logger,
+		subscriptions: make(map[string]map[string]subscription),
+	}
+	return ret, nil
+}
+
+func (ps *pubsub) Subscribe(id, topic string, handler messaging.MessageHandler) error {
+	if id == "" {
+		return ErrEmptyID
+	}
+	if topic == "" {
+		return ErrEmptyTopic
+	}
+	ps.mu.Lock()
+	defer ps.mu.Unlock()
+	// Check topic
+	s, ok := ps.subscriptions[topic]
+	switch ok {
+	case true:
+		// Check topic ID
+		if _, ok := s[id]; ok {
+			return ErrAlreadySubscribed
+		}
+	default:
+		s = make(map[string]subscription)
+		ps.subscriptions[topic] = s
+	}
+
+	queue, err := ps.ch.QueueDeclare(topic, true, true, true, false, nil)
+	if err != nil {
+		return err
+	}
+	if err := ps.ch.QueueBind(queue.Name, queue.Name, exchangeName, false, nil); err != nil {
```

Comment:
Is it safer to use topic instead of `queue.Name`? Is it possible that the broker will change the name passed in QueueDeclare?

Reply:
The broker doesn't change the passed name